feat: add clockSkewInSeconds

stillmatic opened the pull request:
### Discussion

  * See #624 and #625
  * Implements the feedback in  https://github.com/firebase/firebase-admin-python/pull/625#issuecomment-1331197410 - with the suggested API and additional unit/integration tests

### Testing

  * All unit tests pass with `pytest`. 
  * Lint passes.
  * Integration tests are pretty annoying. Not sure how to create an expired token to use, and the session cookies have a minimum expiry time of 300 seconds - the naive sleep 300 means that tests take 5 min to run. Happy to take guidance on improving that.

### API Changes

  * The API changes here were suggested/approved in the linked discussion.

jonathanedey reviewed (changes requested):
Thanks @stillmatic for your contribution and getting this feature moving again.

This looks great! I have attached a few comments to be addressed to fully match the approved internal API proposal so please take a look. 

Additionally, we would like to add validation on the `clock_skew_seconds` to limit it to between `0 and 60` seconds and throw a `Value Error` otherwise. Could you add that here with supporting test cases? Thank you!

stillmatic commented:
I've addressed the feedback in the code change.

Note: this dependency uses `clock_skew_in_seconds` https://github.com/googleapis/google-auth-library-python/blob/7039beb63b8644be748cfc2fc79a2b8b643cda9f/google/oauth2/id_token.py#L112C5-L112C26, So if that gets updated, then this should also be updated. I'm going to assume that moving forward `clock_skew_seconds` is preferred and that you'll handle the discrepancy there.

jonathanedey commented on integration/test_auth.py:
```diff
@@ -577,6 +600,19 @@ def test_verify_id_token_revoked(new_user, api_key):
     claims = auth.verify_id_token(id_token, check_revoked=True)
     assert claims['iat'] * 1000 >= user.tokens_valid_after_timestamp
 
+def test_verify_id_token_tolerance(new_user, api_key):
+    expired_id_token = _sign_in_with_password(new_user.email, 'password', api_key)
```
Should this be `new_user_with_params` here?

I'm guessing it's passing `None` to get an invalid token which is expired by default. Is this the only way to obtain an expired `id_token`?

stillmatic replied:
I'm honestly not sure -- I couldn't find any integration examples with expired id tokens

stillmatic replied:
yep, lints failed on this test, so deleting this actually fixes that as well!

jonathanedey replied:
`test_session_cookies_with_tolerance`, `test_verify_session_cookie_tolerance` and `test_verify_session_cookie_clock_skew_seconds_range` should also be dropped for the same reason. Thank you!

jonathanedey commented on integration/test_auth.py:
```diff
@@ -577,6 +600,19 @@ def test_verify_id_token_revoked(new_user, api_key):
+    time.sleep(1)
+    # Verify the ID token with a tolerance of 0 seconds. This should
+    # raise an exception because the token is expired.
+    with pytest.raises(auth.InvalidIdTokenError) as excinfo:
```
Can we catch `ExpiredIdTokenError` here instead?

jonathanedey commented on integration/test_auth.py:
```diff
@@ -617,6 +653,44 @@ def test_verify_session_cookie_revoked(new_user, api_key):
+    time.sleep(3)
+    # Verify the session cookie with a tolerance of 0 seconds. This should
+    # raise an exception because the cookie is expired.
+    with pytest.raises(auth.InvalidSessionCookieError) as excinfo:
```
Same as above.

jonathanedey reviewed (commented):
A few small questions but otherwise it looks good! 
@lahirumaramba do you have anything to add that we might have missed

Thanks for your concern around the discrepancy. We are aware of the convention there but preferred to default to being more in line with our other Firebase SDKs

lahirumaramba reviewed (approved):
LGTM! Thank you! Please take a look at the failing lint tests

stillmatic commented:
cool - I removed that test and the lint is passing locally now

jonathanedey reviewed (approved):
LGTM! Thank you for your contribution.

stillmatic commented:
thanks for the help reviewing! I've rebased on upstream main and everything should be green.

egilmorez commented on firebase_admin/auth.py:
```diff
@@ -254,6 +256,7 @@ def verify_session_cookie(session_cookie, check_revoked=False, app=None):
         check_revoked: Boolean, if true, checks whether the cookie has been revoked or the
             user disabled (optional).
         app: An App instance (optional).
+        clock_skew_seconds: The number of seconds to tolerate when checking the cookie
```
Suggest period after "cookie."

egilmorez reviewed (approved):
LG with tiny nit, thanks!
